src/category_theory/triangulated/pretriangulated.lean

```diff
@@ -0,0 +1,206 @@
+import category_theory.shift
+import category_theory.preadditive.additive_functor
+import category_theory.triangulated.basic
+import category_theory.triangulated.rotate
+
+/-!
+# Pretriangulated Categories
+
+This file contains the definition of pretriangulated categories and triangulated functors
+between them.
+
+TODO: generalise this to n-angulated categories as in https://arxiv.org/abs/1006.4592
+-/
+
+noncomputable theory
+
+open category_theory
+open category_theory.preadditive
+
+universes v v₀ v₁ v₂ u u₀ u₁ u₂
+
+namespace category_theory.triangulated
+open category_theory.category
+
+/-
+We work in an additive category `C` equipped with an additive shift.
+-/
+variables (C : Type u) [category.{v} C] [has_shift C] [additive_category C]
+[functor.additive (shift C).functor] [functor.additive (shift C).inverse]
+
+/--
+An additive category `C` with an additive shift, and a class of "distinguished triangles"
+relative to that shift is called pretriangulated if the following hold:
+* Any triangle that is isomorphic to a distinguished triangle is also distinguished.
+* Any triangle of the form `(X,X,0,id,0,0)` is distinguished.
+* For any morphism `f: X ⟶ Y` there exists a distinguished triangle of the form `(X,Y,Z,f,g,h)`.
+* The triangle `(X,Y,Z,f,g,h)` is distinguished if and only if `(Y,Z,X⟦1⟧,g,h,-f⟦1⟧)` is.
+* Given a commutative diagram:
+```
+        f       g       h
+    X  ───> Y  ───> Z  ───> X⟦1⟧
+    │       │                │
+    │a      │b               │a⟦1⟧'
+    V       V                V
+    X' ───> Y' ───> Z' ───> X'⟦1⟧
+        f'      g'      h'
+```
+  whose rows are distinguished triangles, there exists a morphism `c: Z ⟶ Z'` such that `(a,b,c)`
+  is a triangle morphism.
+See https://stacks.math.columbia.edu/tag/0145
+-/
+class pretriangulated :=
+(distinguished_triangles : set(triangle C))
+(isomorphic_distinguished : Π (T₁ ∈ distinguished_triangles) (T₂ : triangle C) (T₁ ≅ T₂),
+  T₂ ∈ distinguished_triangles)
+(contractible_distinguished : Π (X : C), (contractible_triangle C X) ∈ distinguished_triangles)
+(distinguished_cocone_triangle : Π (X:C) (Y:C) (f: X ⟶ Y), (∃ (Z : C) (g : Y ⟶ Z) (h : Z ⟶ X⟦1⟧),
+  ({ obj₁ := X, obj₂ := Y, obj₃ := Z, mor₁ := f, mor₂ := g, mor₃ := h} : triangle C) ∈
+    distinguished_triangles))
+(rotate_distinguished_triangle : Π (T : triangle C),
+  T ∈ distinguished_triangles ↔ T.rotate C ∈ distinguished_triangles)
+(complete_distinguished_triangle_morphism : Π (T₁ T₂ : triangle C)
+  (h₁ :T₁ ∈ distinguished_triangles) (h₂ :T₂ ∈ distinguished_triangles) (a : T₁.obj₁ ⟶ T₂.obj₁)
+  (b : T₁.obj₂ ⟶ T₂.obj₂) (comm₁ : T₁.mor₁ ≫ b = a ≫ T₂.mor₁),
+  (∃ (c : T₁.obj₃ ⟶ T₂.obj₃), (T₁.mor₂ ≫ c = b ≫ T₂.mor₂) ∧ (T₁.mor₃ ≫ a⟦1⟧' = c ≫ T₂.mor₃) ))
+
+/--
+Given any distinguished triangle `T`, then we know `T.rotate C` is also distinguished.
+-/
+lemma rot_of_dist_triangle [CT : pretriangulated C] (T ∈ CT.distinguished_triangles) :
+  (T.rotate C ∈ CT.distinguished_triangles) :=
```

Comment:
Shouldn't you be writing `distinguished_triangles C` rather than `CT.distinguished_triangles`? If that isn't working something must be set up wrong. (Maybe try `(distinguished_triangles [] : set (triangle C))`?

Reply:
(Generally, if you find yourself having to name the instance variables it's usually a bad sign.)